src/v/cloud_topics/level_one/metastore.h

```diff
@@ -109,6 +110,96 @@ class metastore {
+            bool operator==(const cleaned_range&) const = default;
+            kafka::offset base_offset;
+            kafka::offset last_offset;
+            std::optional<model::timestamp> cleaned_with_tombstones_at;
+        };
+        // A range indicating that the data's keys have been fully deduplicated
+        // from the start of the log.
+        std::optional<cleaned_range> new_cleaned_range;
+
+        // Ranges of offsets that previously had tombstones, that have been
+        // removed.
+        chunked_vector<cleaned_range> removed_tombstones_ranges;
+    };
+    using compaction_map_t
+      = chunked_hash_map<model::topic_id_partition, compaction_update>;
+    struct compaction_offsets_response {
+        // Offset ranges whose keys have not been fully deduplicated from the
+        // start of the log.
+        offset_interval_set dirty_ranges;
+
+        // The set of offset ranges that contain tombstones whose keys have
+        // been cleaned long enough to be eligible for tombstone removal.
+        //
+        // A compaction method, when iterating over a tombstone record, may
+        // consult this to determine if the tombstone should be removed.
+        offset_interval_set removable_tombstone_ranges;
+    };
+    // Similar to replace_objects(), but with additional constraints based on
+    // compaction metadata.
+    virtual ss::future<std::expected<void, errc>> compact_objects(
+      const chunked_vector<object_metadata>&, const compaction_map_t&)
+      = 0;
+
+    // Returns metadata required to determine what to compact for the given
+    // partition. Below is pseudocode for sample usage:
+    //
+    // offsets = co_await metastore.get_compaction_offsets( \
+    //   partition, tombstone_removal_upper_bound_ts);
+    //
+    // key_offset_map m;
+    // cleaned_ranges new_cleaned_ranges;
+    // cleaned_ranges removed_tombstones_ranges;
+    //
+    // # Build an offset map based on the dirty ranges.
+    // for dirty_range in offsets.dirty_ranges:
+    //     reader = log.reader(dirty_range.base, dirty_range.last)
+    //     co_await m.add_latest_offset_per_key(reader)
+    //
+    //     cleaned_range r(...offset range that was actually read...);
+    //     if ...reader witnessed tombstones...:
+    //         cleaned_range.cleaned_with_tombstones_at = now()
+    //
+    //     new_cleaned_ranges.insert(cleaned_range)
+    //
+    // # Determine what ranges to remove tombstones from.
+    // removed_tombstones_ranges = \
+    //   ...offsets.removable_tombstone_ranges that fall below m.max_offset()...
+    //
+    // # This operation deduplicates based on the offset map and removes
+    // # tombstones in the given ranges, up to the max indexed by the map.
+    // objects = co_await log.compact( \
+    //   m.max_offset(), m, removed_tombstones_ranges)
+    //
+    // co_await metastore.compact_objects( \
+    //   objects, {new_cleaned_ranges, removed_tombstones_ranges})
```

Comment:
nit: you changed this to be a `map_t` of partitions, would be good to show how that factors in.